File: src/lib/snooze.ts
import { invoke, isTauri } from "@tauri-apps/api/core";

export interface SnoozedMessage {
  id: string;
  account_id: string;
  folder: string | null;
  uid: number | null;
  pop3_uidl: string | null;
  subject: string | null;
  sender: string | null;
  snooze_until: string;
  created_at: string;
}

export interface SnoozeArgs {
  accountId: string;
  folder?: string | null;
  uid?: number | null;
  pop3Uidl?: string | null;
  subject?: string | null;
  sender?: string | null;
  snoozeUntil: string; // RFC 3339 / ISO 8601
}

export function snoozeMessage(args: SnoozeArgs): Promise<string> {
  if (!isTauri()) return Promise.resolve("mock-snooze-id");
  return invoke("snooze_message", {
    accountId: args.accountId,
    folder: args.folder ?? null,
    uid: args.uid ?? null,
    pop3Uidl: args.pop3Uidl ?? null,
    subject: args.subject ?? null,
    sender: args.sender ?? null,
    snoozeUntil: args.snoozeUntil,
  });
}

export function cancelSnooze(id: string): Promise<void> {
  if (!isTauri()) return Promise.resolve();
  return invoke("cancel_snooze", { id });
}

export function listDueSnoozed(): Promise<SnoozedMessage[]> {
  if (!isTauri()) return Promise.resolve([]);
  return invoke("list_due_snoozed");
}

export function listAllSnoozed(accountId: string): Promise<SnoozedMessage[]> {
  if (!isTauri()) return Promise.resolve([]);
  return invoke("list_all_snoozed", { accountId });
}

File: src-tauri/src/snooze.rs
use crate::cache::{self, SnoozedMessage};

fn generate_id() -> String {
    use rand::RngCore;
    let mut bytes = [0u8; 16];
    rand::rngs::OsRng.fill_bytes(&mut bytes);
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[tauri::command]
pub async fn snooze_message(
    account_id: String,
    folder: Option<String>,
    uid: Option<i64>,
    pop3_uidl: Option<String>,
    subject: Option<String>,
    sender: Option<String>,
    snooze_until: String,
) -> Result<String, String> {
    let id = generate_id();
    let conn = cache::open()?;
    cache::insert_snooze(
        &conn,
        &id,
        &account_id,
        folder.as_deref(),
        uid,
        pop3_uidl.as_deref(),
        subject.as_deref(),
        sender.as_deref(),
        &snooze_until,
    )?;
    Ok(id)
}

#[tauri::command]
pub async fn cancel_snooze(id: String) -> Result<(), String> {
    let conn = cache::open()?;
    cache::delete_snooze(&conn, &id)
}

#[tauri::command]
pub async fn list_due_snoozed() -> Result<Vec<SnoozedMessage>, String> {
    let conn = cache::open()?;
    cache::list_due_snoozed(&conn)
}

#[tauri::command]
pub async fn list_all_snoozed(account_id: String) -> Result<Vec<SnoozedMessage>, String> {
    let conn = cache::open()?;
    cache::list_all_snoozed(&conn, &account_id)
}
